% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
\documentclass[11pt]{article}
%% Set my margins
\setlength{\oddsidemargin}{0.0truein}
\setlength{\evensidemargin}{0.0truein}
\setlength{\textwidth}{6.5truein}
\setlength{\topmargin}{0.0truein}
\setlength{\textheight}{9.0truein}
\setlength{\headsep}{0.0truein}
\setlength{\headheight}{0.0truein}
\setlength{\topskip}{0pt}
%% End of margins

% \usepackage{subfigure}

%%\pagestyle{myheadings}
%%\markboth{$Date$\hfil$Revision$}{\thepage}
\usepackage[pdftex,
bookmarks,
bookmarksopen,
pdfauthor={Jingsi Ming},
pdftitle={FIRM Vignette}]
{hyperref}

\title{`\texttt{FIRM}' Package for flexible integration of scRNA-seq data}

\date{\today}

\SweaveOpts{engine=R, echo=TRUE, pdf=TRUE}

\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle

\section{Overview}

This vignette provides an introduction to the `\texttt{FIRM}' package.
R package `\texttt{FIRM}' implements FIRM, an algorithm for flexible integration of heterogeneous scRNA-seq datasets across multiple tissue types, platforms and experimental batches.

The FIRM package can be loaded with the command:

<<preliminaries,echo=FALSE,results=hide>>=
options(prompt = "R> ")
@

<<prelim>>=
library("FIRM")
@

Also load required package:
<<prelim>>=
library(Seurat)
library(RANN)
@

\section{Workflow}\label{workflow}

In this vignette, we use the ExampleData in the package which is the aorta data from Tabula Muris generated using Smart-seq2 (SS2) and 10X. The `\texttt{ExampleData\$SS2}' and `\texttt{ExampleData\$tenx}' provide the gene expression matrices $\mathbf{X}$ for SS2 and 10X respectively, where $X_{ij}$ is the number of reads (for SS2) and unique molecular identified (UMI, for 10X). We only considered the cells with annotations in the original study. The cell type annotations for SS2 and 10X datasets are provided in `\texttt{ExampleData\$meta\_SS2}' and `\texttt{ExampleData\$meta\_tenx}'.

<<prelim>>=
data("ExampleData")
@

<<prelim, echo=FALSE, results=hide>>=
dim(SS2)
@

<<prelim>>=
dim(SS2)
length(meta_SS2)
dim(tenx)
length(meta_tenx)
@

\subsection{Data preprocessing}
We performed the standard preprocessing workflow to prepare the scaled data for integration.
<<preproc>>=
prep_SS2 <- prep_data(SS2)
Dataset1 <- prep_SS2$Dataset
hvg1 <- prep_SS2$hvg

prep_tenx <- prep_data(tenx)
Dataset2 <- prep_tenx$Dataset
hvg2 <- prep_tenx$hvg
@

\subsection{Integration using FIRM}
The integrated data is provided in `\texttt{Dataset}'.
<<FIRM, warning=FALSE, result=hide, message=FALSE>>=
dims <- 15
coreNum <- 4
Dataset <- FIRM(Dataset1, Dataset2, all_genes = FALSE, 
                hvg1, hvg2, dims = dims, coreNum = coreNum)
@

<<FIRM, warning=FALSE, result=hide>>=
dim(Dataset)
@

For downstream analysis which focuses on the highly variable genes shared by the datasets, such as dimension reduction, clustering and visualization, we set `\texttt{all\_genes = FALSE}' to improve the computational efficiency and memory usage, which is the default setting in `\texttt{FIRM}' function.
For downstream analysis which requires full gene expression profiles, we set `\texttt{all\_genes = TRUE}' to obtain the harmonized data for all genes.

<<FIRM, warning=FALSE, result=hide, message=FALSE>>=
Dataset <- FIRM(Dataset1, Dataset2, all_genes = TRUE, 
                hvg1, hvg2, dims = dims, coreNum = coreNum)
@

<<FIRM, warning=FALSE, result=hide>>=
dim(Dataset)
@

\subsection{Downstream analysis}
We can use the integrated data for downstream analysis, such as PCA and visaulization. For example, we can create a Seurat object.
<<create_seurat, warning=FALSE, message=FALSE>>=
# counts
integrated_counts <- matrix(0, nrow(Dataset), ncol(Dataset))
rownames(integrated_counts) <- rownames(Dataset)
colnames(integrated_counts) <- colnames(Dataset)
integrated_counts[rownames(SS2), colnames(SS2)] <- as.matrix(SS2)
integrated_counts[rownames(tenx), colnames(tenx)] <- as.matrix(tenx)

# create Seurat object
integrated <- CreateSeuratObject(integrated_counts)

# normalization
integrated <- NormalizeData(integrated)

# put the centered integrated data in scaled data
integrated[["RNA"]]@scale.data <- Dataset - rowMeans(Dataset)

# add meta data
integrated <- AddMetaData(integrated, metadata = c(meta_SS2, meta_tenx),
                          col.name = "annotation")
integrated <- AddMetaData(integrated,
                          metadata = c(rep("SS2", ncol(SS2)), rep("10X", ncol(tenx))),
                          col.name = "dataset")

# hvg for PCA and visualization
hvg <- intersect(hvg1, hvg2)

# PCA
integrated <- RunPCA(integrated, features = hvg, npcs = dims)

# UMAP
integrated <- RunUMAP(integrated, reduction = "pca", dims = 1:dims,
                      umap.method = 'umap-learn', metric = "correlation")
@

<<fig=TRUE>>=
DimPlot(integrated, reduction = "umap", group.by = "dataset")
@

<<fig=TRUE>>=
library(ggplot2)
DimPlot(integrated, reduction = "umap", group.by = "annotation") +
  theme(legend.position = "top")
@

\begin{thebibliography}{99}
\bibitem{Ming} Jingsi Ming, Zhixiang Lin, Jia Zhao, Xiang Wan, The Tabula Microcebus Consortium, Can Yang, Angela Ruohao Wu, FIRM: Flexible Integration of single-cell RNA-sequencing data for large-scale Multi-tissue cell atlas datasets. Briefings in Bioinformatics (2022). https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbac167/6585621

\bibitem{Schaum} Schaum, N. et al. Single-cell transcriptomics of 20 mouse organs creates a Tabula Muris. Nature 562, 367â€“372 (2018).

\bibitem{Stuart} Stuart, T. et al. Comprehensive Integration of Single-Cell Data. Cell 177, 1888-1902.e21 (2019).

\end{thebibliography}

\end{document}
